server/sonar-server/src/main/java/org/sonar/server/computation/step/CustomMeasuresCopyStep.java

```diff
@@ -0,0 +1,116 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.step;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.List;
+import org.apache.commons.lang.math.NumberUtils;
+import org.sonar.core.measure.custom.db.CustomMeasureDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.measure.Measure;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.computation.metric.Metric;
+import org.sonar.server.computation.metric.MetricRepository;
+import org.sonar.server.db.DbClient;
+
+public class CustomMeasuresCopyStep implements ComputationStep {
+
+  private final TreeRootHolder treeRootHolder;
+  private final DbClient dbClient;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  public CustomMeasuresCopyStep(TreeRootHolder treeRootHolder, DbClient dbClient,
+    MetricRepository metricRepository, MeasureRepository measureRepository) {
+    this.treeRootHolder = treeRootHolder;
+    this.dbClient = dbClient;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+  }
+
+  @Override
+  public void execute() {
+    new DepthTraversalTypeAwareVisitor(Component.Type.FILE, DepthTraversalTypeAwareVisitor.Order.PRE_ORDER) {
+      @Override
+      public void visitAny(Component component) {
+        copy(component);
+      }
+    }.visit(treeRootHolder.getRoot());
+  }
+
+  private void copy(Component component) {
+    for (CustomMeasureDto dto : loadCustomMeasures(component)) {
+      Metric metric = metricRepository.getById(dto.getMetricId());
+      // TODO verify that deletion of custom metrics correctly delete related measures,
+      // else metric is not found and an exception is raised
+      Measure measure = dtoToMeasure(dto, metric);
+      measureRepository.add(component, metric, measure);
+    }
+  }
+
+  private List<CustomMeasureDto> loadCustomMeasures(Component component) {
+    DbSession session = dbClient.openSession(false);
+    try {
+      return dbClient.customMeasureDao().selectByComponentUuid(session, component.getUuid());
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  @VisibleForTesting
+  static Measure dtoToMeasure(CustomMeasureDto dto, Metric metric) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "dtoToMeasure" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)